### ISSUE TITLE
Are the next N elements set?

### ISSUE THREAD

**Issue by richardstartin** [RoaringBitmap/RoaringBitmap#213]
I am using `Container.forEach` for aggregations over circuits in a toy project. The code isn't as fast as it could be because the lambda `IntConsumer` typically won't inline. Even simple numeric code doesn't vectorise in this case. In this context I have a page consisting of, say, a `double[]`. If I know the next 128 bits are set, I can iterate directly on the `double[]` and the code is likely to vectorise. 
```java
boolean containsRange(short minimum, short supremum);
```
Is this semantic available? Should it be implemented? I am aware that `Container`s are not considered public API.

**Comment by lemire** [comment]
The highest performance should probably be achievable by using "container-aware" iterators. That is, instead of calling your function for each integer, we call it for each container. Of course, your function has to "know" about container types. And then we have the "public API" problem... 

I think we have ranged add, remove and flip... but not "contains". It would make sense to add it.

Is there a way for you to rewrite your code so that you are not so reliant on forEach?



It would be useful if you were to contribute a benchmark so we can back with numbers the assertion "the code isn't as fast as it could be". That is, what are we talking about? A 10x inefficiency or a 10% inefficiency? I would also urge you to benchmark different approaches including using iterators instead of forEach.

**Comment by richardstartin** [comment]
OK. The crunch is that `forEach` style is not a countable iteration strategy, will only inline in the best case (probably in a microbenchmark), and this effect is more observable for compressed data. You can do much better by iterating over 256 bit chunks. Once you have realistic programmatic usage patterns, the difference is stark, but it is sensitive on particular data sets and sort orders. Here's a brief analysis of one of my own benchmarks. 

Consider this pair of benchmarks. They both compute the spread between two series of observations. The first one uses `forEach` and the second one uses some C like code, with a "representative cost" of doing a `containsRange` check. The representative cost is too low, and the benchmark only measures the case where I have seen performance problems. 

```java
  private static final short ZERO = 0;

  private Container rc;
  private final double[] x = new double[1 << 16];
  private final double[] y = new double[1 << 16];
  private final double[] z = new double[1 << 16];


  @Setup(Level.Trial)
  public void setup() {
    IntStream.range(0, 1 << 16).forEach(i -> x[i] = ThreadLocalRandom.current().nextDouble());
    IntStream.range(0, 1 << 16).forEach(i -> y[i] = x[i] + ThreadLocalRandom.current().nextDouble());;
    rc = RunContainer.full();
  }

  @Benchmark
  public void monomorphicForEach(Blackhole bh) {
    rc.forEach(ZERO, i -> z[i] = x[i] - y[i]);
    bh.consume(z);
  }

  @Benchmark
  public void bulk1(Blackhole bh) {
    final int cardinality = rc.getCardinality();
    final int offset = 0;
    final int stride = 1 << 8; // magic point where vectorisation kicks in
    int card = cardinality;
    int iteration = 0;
    while (card > 0) {
      int min = iteration * offset;
      int sup = (iteration * offset) + stride;
      // factor in some kind of cost similar to doing .containsRange
      if (rc.contains((short) min) && rc.contains((short) (sup - 1))) {
        for (int i = min; i < sup; ++i) {
          z[i] = x[i] - y[i];
        }
        card -= stride;
      }
      ++iteration;
    }
    bh.consume(z);
  }
```
Consider the worst possible case for `forEach` - a full `RunContainer`. The `IntConsumer` gets inlined because there's only one of them, but the loop isn't countable, so it isn't unrolled and isn't vectorised. You see the `IntConsumer` is inlined to this:

```asm
  0x000001ab12161fa3: vsubsd  xmm1,xmm1,mmword ptr [rax+r9*8+10h]
```

In the bulk case you see better code (unrolled, vectorised):

```asm
  0x000001d3313e1631: vmovdqu ymm2,ymmword ptr [rbx+r11*8+10h]
  0x000001d3313e1638: vsubpd  ymm2,ymm2,ymmword ptr [rdi+r11*8+10h]
  0x000001d3313e163f: vmovdqu ymmword ptr [rsi+r11*8+10h],ymm2
  0x000001d3313e1646: vmovdqu ymm2,ymmword ptr [rbx+r11*8+30h]
  0x000001d3313e164d: vsubpd  ymm2,ymm2,ymmword ptr [rdi+r11*8+30h]
  0x000001d3313e1654: vmovdqu ymmword ptr [rsi+r11*8+30h],ymm2
  0x000001d3313e165b: vmovdqu ymm2,ymmword ptr [rbx+r11*8+50h]
  0x000001d3313e1662: vsubpd  ymm2,ymm2,ymmword ptr [rdi+r11*8+50h]
  0x000001d3313e1669: vmovdqu ymmword ptr [rsi+r11*8+50h],ymm2
  0x000001d3313e1670: vmovdqu ymm2,ymmword ptr [rbx+r11*8+70h]
  0x000001d3313e1677: vsubpd  ymm2,ymm2,ymmword ptr [rdi+r11*8+70h]
  0x000001d3313e167e: vmovdqu ymmword ptr [rsi+r11*8+70h],ymm2
  0x000001d3313e1685: vmovdqu ymm2,ymmword ptr [rbx+r11*8+90h]
  0x000001d3313e168f: vsubpd  ymm2,ymm2,ymmword ptr [rdi+r11*8+90h]
  0x000001d3313e1699: vmovdqu ymmword ptr [rsi+r11*8+90h],ymm2
  0x000001d3313e16a3: vmovdqu ymm2,ymmword ptr [rbx+r11*8+0b0h]
  0x000001d3313e16ad: vsubpd  ymm2,ymm2,ymmword ptr [rdi+r11*8+0b0h]
  0x000001d3313e16b7: vmovdqu ymmword ptr [rsi+r11*8+0b0h],ymm2
  0x000001d3313e16c1: vmovdqu ymm2,ymmword ptr [rbx+r11*8+0d0h]
  0x000001d3313e16cb: vsubpd  ymm2,ymm2,ymmword ptr [rdi+r11*8+0d0h]
  0x000001d3313e16d5: vmovdqu ymmword ptr [rsi+r11*8+0d0h],ymm2
  0x000001d3313e16df: vmovdqu ymm2,ymmword ptr [rbx+r11*8+0f0h]
  0x000001d3313e16e9: vsubpd  ymm2,ymm2,ymmword ptr [rdi+r11*8+0f0h]
  0x000001d3313e16f3: vmovdqu ymmword ptr [rsi+r11*8+0f0h],ymm2
```

And it shows with a 4x throughput difference (`vsubpd` couldn't do any more with 256 bit registers):

| Benchmark                                                          |   Mode  | Cnt   |  Score + Error   |   Units   |
|-----------------------------------------------------|----------|------:|-----------------:|----------|
| VectorisationBenchmark.bulk1                            |  thrpt    | 10     | 99.239 ± 2.207 | ops/ms  |
| VectorisationBenchmark.monomorphicForEach |  thrpt    | 10      | 25.434 ± 0.321 |  ops/ms |

Once you have enough `IntConsumers` to disable inlining, it gets worse:

```java
  @Benchmark
  public void megamorphicForEach(Blackhole bh) {
    rc.forEach(ZERO, i -> z[i] = x[i] - y[i]);
    bh.consume(z);
    rc.forEach(ZERO, i -> z[i] = z[i] - x[i] * y[i]);
    bh.consume(z);
    rc.forEach(ZERO, i -> z[i] = z[i] + x[i] * y[i]);
    bh.consume(z);
  }

@Benchmark
  public void bulk3(Blackhole bh) {
    final int cardinality = rc.getCardinality();
    final int offset = 0;
    final int stride = 1 << 8; // magic point where vectorisation kicks in
    int card = cardinality;
    int iteration = 0;
    while (card > 0) {
      int min = iteration * offset;
      int sup = (iteration * offset) + stride;
      // factor in some kind of cost similar to doing .containsRange
      if (rc.contains((short) min) && rc.contains((short) (sup - 1))) {
        for (int i = min; i < sup; ++i) {
          z[i] = x[i] - y[i];
        }
        card -= stride;
      }
      ++iteration;
    }
    bh.consume(z);
    card = cardinality;
    iteration = 0;
    while (card > 0) {
      int min = iteration * offset;
      int sup = (iteration * offset) + stride;
      // factor in some kind of cost similar to doing .containsRange
      if (rc.contains((short) min) && rc.contains((short) (sup - 1))) {
        for (int i = min; i < sup; ++i) {
          z[i] = z[i] - x[i] * y[i];
        }
        card -= stride;
      }
      ++iteration;
    }
    bh.consume(z);
    card = cardinality;
    iteration = 0;
    while (card > 0) {
      int min = iteration * offset;
      int sup = (iteration * offset) + stride;
      // factor in some kind of cost similar to doing .containsRange
      if (rc.contains((short) min) && rc.contains((short) (sup - 1))) {
        for (int i = min; i < sup; ++i) {
          z[i] = z[i] + x[i] * y[i];
        }
        card -= stride;
      }
      ++iteration;
    }
    bh.consume(z);
  }
```

All of the C like code is unrolled and vectorised. Whereas the `forEach` consumer is now an invocation of a Lambda form:

```asm
  0x000001bae98d85f9: jmp     1bae1edb800h      ;...e9
                                                ;...02
                                                ;...32
                                                ;...60
                                                ;...f8
                                                ;*putfield arg$1 {reexecute=0 rethrow=0 return_oop=0}
                                                ; - com.openkappa.splitmap.VectorisationBenchmark$$Lambda$42/1529283420::<init>@6
                                                ; - com.openkappa.splitmap.VectorisationBenchmark$$Lambda$42/1529283420::get$Lambda@5
                                                ; - java.lang.invoke.DirectMethodHandle$Holder::invokeStatic@10
                                                ; - java.lang.invoke.LambdaForm$MH/707610042::linkToTargetMethod@5
                                                ; - com.openkappa.splitmap.VectorisationBenchmark::megamorphicForEach@50 (line 45)
                                                ;   {runtime_call _rethrow_Java}

```
| Benchmark                                                          |   Mode  | Cnt   |  Score + Error   |   Units   |
|-----------------------------------------------------|----------|------:|-----------------:|----------|
| VectorisationBenchmark.bulk3                            |  thrpt    | 10     | 27.185 ± 0.242 | ops/ms  |
| VectorisationBenchmark.megamorphicForEach |  thrpt    | 10      | 2.778 ± 4.485   |  ops/ms |

This is now a 10x difference in throughput.

But there is a cost to handling the other branch not captured here (when you need to iterate over noncontiguous subranges, I would have to implement an `iterateBetween` strategy to measure its performance), and for virtually empty `RunContainer`s the difference is negligible. So it depends on the data whether this is worth doing.

**Comment by lemire** [comment]
Thanks for documenting the issue. I think I understand the need for something better... but I am now thinking about how to solve the issue in a practical manner for most users without exposing more of our internals than needed.

I am not sure how containsRange solves the problem in this case. I mean... it might solve a performance issue if you are hacking deep in the code... but it does little for Joe or Jill, the average user.

**Comment by richardstartin** [comment]
Rather than hacking deep in the code, I am using `Container` in a different data structure outside of this project, because I am unaware of a better 16 bit bitset implementation. Would you discourage this? In any case, the idea is if there is a contiguous range, interaction with `Container` can be replaced with a for loop which doesn't touch the `Container`.

**Comment by lemire** [comment]
Ok. I think it is clearer now.

So we would like to have a containsRange method. This is a good idea.

**Comment by richardstartin** [comment]
At the same time, I see it as being slightly problematic, because if I were to implement it, it would be for my own benefit and it wouldn't be for a use case within this library. It's also only half the solution to my problem: there needs to be a way to iterate over an N bit range too, in case the range isn't dense. Still, I think there must be faster ways to extract bits from bitmaps than are currently available. It also occurred to me that words rather than bits could be iterated over.

```java
interface WordIterator {
   boolean hasNext();
   // returns the next 64 bits packed into a long
   long nextAsLong();
}
```
I can implement my use case from this:


```java
WordIterator wi = container.getWordIterator();
for (int i = 0; i + 4 < 1024; i += 4) {
    long w1 = wi.nextAsLong();
    long w2 = wi.nextAsLong();
    long w3 = wi.nextAsLong();
    long w4 = wi.nextAsLong();
    if (w1 + w2 + w3 + w4 == -4L) { // contiguous
       for (int j = 0; j < 256; ++j) {
         // fast vectorised loop
       }
   }
   ...
   // or if I have an invertible operation
   if (Long.bitCount(w1) + Long.bitCount(w2) + Long.bitCount(w3) + Long.bitCount(w4) >= threshold) { // contiguous
       for (int j = 0; j < 256; ++j) {
         // fast vectorised loop
       }
      // repair 
       forEachUnsetBit(w1, inverseOp);
       forEachUnsetBit(w2, inverseOp);
       forEachUnsetBit(w3, inverseOp);
       forEachUnsetBit(w4, inverseOp);
   }
}
```

This is extremely ugly, even by my standards, but the point is, a lot can be done _outside_ this library with this functionality. If this is generally useful, maybe it's worth implementing.

**Comment by richardstartin** [comment]
It was possible to double the throughput of vectorisable calling code when iterating over bitmaps composed of all three container types by implementing `containsRange` with a series of precomputed `RunContainer`s containing disjoint 256 bit masks, and calling `PeekableShortIterator.advanceIfNeeded` whenever the fast path is taken.

**Comment by lemire** [comment]
Noted.

**Comment by richardstartin** [comment]
Implemented in #237
